### PR TITLE
Add pmwcs3 capacitive soil moisture & temperature sensor component

### PR DESCRIPTION
**Pull request by SeByDocKy**
# What does this implement/fix?

Add the pmwcs3 capacitive soil moisture, temperature sensor with I2C communication.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2791<esphome-docs PR number goes here>

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
sensor:
  - platform: pmwcs3
    i2c_id: i2c_bus
    e25:
      name: pmwcs3_e25
    ec:
      name: pmwcs3_ec
    temperature:
      name: pmwcs3_temperature
    vwc:
      name: pmwcs3_vwc

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/pmwcs3/pmwcs3.cpp
Comment (jesserockz) [inline esphome/components/pmwcs3/pmwcs3.cpp:1]: Please remove the comments with `NOLINT delay`

File: esphome/components/pmwcs3/pmwcs3.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+void PMWCS3Component::set_air_calibration() {
+  if (!this->write_bytes(PMWCS3_REG_CALIBRATE_AIR, nullptr, 0)) {
+    this->status_set_warning();
+    ESP_LOGW(TAG, "couldn't start air calibration");
+    return;
+  }
+  ESP_LOGW(TAG, "Start air calibration during the next 300s");
+  // NOLINT  delay(300000);
+  ESP_LOGW(TAG, "Air calibration finished");
```
Comment (jesserockz) [inline esphome/components/pmwcs3/pmwcs3.cpp]: If this process takes 300s, you should set a flag that stops the component from trying to read data until it is done.

You can use `set_timeout` to run code 300s later to set the flag back to false.

File: esphome/components/pmwcs3/pmwcs3.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+static const uint8_t PMWCS3_REG_READ_VWC = 0x05;
+static const uint8_t PMWCS3_REG_CALIBRATE_AIR = 0x06;
+static const uint8_t PMWCS3_REG_CALIBRATE_WATER = 0x07;
+static const uint8_t PMWCS3_SET_I2C_ADDRESS = 0x08;
+static const uint8_t PMWCS3_REG_GET_DATA = 0x09;
+static const uint8_t PMWCS3_REG_CALIBRATE_EC = 0x10;
+static const uint8_t PMWCS3_REG_CAP = 0x0A;
+static const uint8_t PMWCS3_REG_RES = 0x0B;
+static const uint8_t PMWCS3_REG_RC = 0x0C;
+static const uint8_t PMWCS3_REG_RT = 0x0D;
```
Comment (jesserockz) [inline esphome/components/pmwcs3/pmwcs3.h]: These can be moved into the cpp file

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/pmwcs3/pmwcs3.cpp (hunk 1 above)
Comment (SeByDocKy) [reply esphome/components/pmwcs3/pmwcs3.cpp]: > set_timeout

In fact if the read_register is not written (as in the beginning of the read_data() method), no new data are avalable on the I2C bus. Anyway I will look to set_timeout method :)

File: esphome/components/pmwcs3/pmwcs3.cpp [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  LOG_I2C_DEVICE(this);
+  if (this->is_failed()) {
+    ESP_LOGE(TAG, "Communication with PMWCS3 failed!");
+  }
+  ESP_LOGI(TAG, "%s", this->is_failed() ? "FAILED" : "OK");
+
+  LOG_UPDATE_INTERVAL(this);
+  LOG_SENSOR("  ", "e25", this->e25_sensor_);
+  LOG_SENSOR("  ", "ec", this->ec_sensor_);
+  LOG_SENSOR("  ", "temperatue", this->temperature_sensor_);
```
Comment (sebcaps) [inline esphome/components/pmwcs3/pmwcs3.cpp]: small typo :-) temperatuRe

Comment (SeByDocKy) [reply esphome/components/pmwcs3/pmwcs3.cpp]: Dammned you see everything :)

File: esphome/components/pmwcs3/pmwcs3.cpp [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+static const uint8_t PMWCS3_REG_GET_DATA = 0x09;
+static const uint8_t PMWCS3_REG_CALIBRATE_EC = 0x10;
+static const uint8_t PMWCS3_REG_CAP = 0x0A;
+static const uint8_t PMWCS3_REG_RES = 0x0B;
+static const uint8_t PMWCS3_REG_RC = 0x0C;
+static const uint8_t PMWCS3_REG_RT = 0x0D;
+
+static const char *const TAG = "pmwcs3";
+
+void PMWCS3Component::change_i2c_address(uint8_t newaddress) {
```
Comment (sebcaps) [inline esphome/components/pmwcs3/pmwcs3.cpp]: can't see how / when it's called ?
For future use ?

File: esphome/components/pmwcs3/pmwcs3.cpp [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+  if (!this->write_byte(PMWCS3_SET_I2C_ADDRESS, newaddress)) {
+    this->status_set_warning();
+    ESP_LOGW(TAG, "couldn't write the new I2C address %d", newaddress);
+    return;
+  }
+  ESP_LOGVV(TAG, "new I2C address %d done", newaddress);
+}
+
+void PMWCS3Component::set_air_calibration() {
```
Comment (sebcaps) [inline esphome/components/pmwcs3/pmwcs3.cpp]: idem : how to use it ?
For a future use ?

File: esphome/components/pmwcs3/pmwcs3.cpp [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+  if (!this->write_bytes(PMWCS3_REG_CALIBRATE_AIR, nullptr, 0)) {
+    this->status_set_warning();
+    ESP_LOGW(TAG, "couldn't start air calibration");
+    return;
+  }
+  ESP_LOGW(TAG, "Start air calibration during the next 300s");
+}
+void PMWCS3Component::set_water_calibration() {
```
Comment (sebcaps) [inline esphome/components/pmwcs3/pmwcs3.cpp]: For future use ?

**Comment by sebcaps** [comment]
@SeByDocKy you requested review I did it ;-).
Initally  had a look to this PR since I know your youtube channel and permit myself to let a comment.
Can't test it, since I don't have this device (yet), but may give a help to use the three unused function.
Looking forward to see your next video.

**Comment by SeByDocKy** [comment]
> @SeByDocKy you requested review I did it ;-). Initally had a look to this PR since I know your youtube channel and permit myself to let a comment. Can't test it, since I don't have this device (yet), but may give a help to use the three unused function. Looking forward to see your next video.

Yes as you noticed, they are 3 methods not documented coz even in the original arduino lib there are no example how to use them. Anyway I more or less understood how. My proposal was(is) to use these methods in dedicaced yaml model : changing the I2C adress, calibrating air and water measures. To proceed, I propose to definie a lambda in the on_boot section and run these methods like

````
esphome:
  name: ${name}
  
  # on_boot:
  #   priority: -100
  #   then:

  #     ######### Uncomment the next lamnda section to change I2C address, then replace the new I2C address in the sensor part ########

  #     - lambda: |-
  #         esphome::pmwcs3::PMWCS3Component *sen =  ${name}_pmwc3;
  #         sen->change_i2c_address(0x65);

  #     #########  Uncomment the next lamnda section to perform the air/dry soil calibration   ##########
      
  #     - lambda: |-
  #         esphome::pmwcs3::PMWCS3Component *sen =  ${name}_pmwc3;
  #         sen->set_air_calibration();

  #     #########  Uncomment the next lamnda section to perform the water/humid soil calibration   ##########
      
  #     - lambda: |-
  #         esphome::pmwcs3::PMWCS3Component *sen =  ${name}_pmwc3;
  #         sen->set_water_calibration();```


When I asked if I was correct to propose this method (via the on_boot), I was said should be avoided (even if for the pzem ac, to change the modbus address, the same trick is used)... this is why I didn't include some how to use these 3 methods in my original PR

**Comment by sebcaps** [comment]
Hello,
Definitely device doc is very light, but arduino library quite explicit. 
From a style perspective, it's a pity to declare functions not being able to use them. If usage of these lambda in boot is discouraged I would stick to the advice.
Setting i2c adress via code is somehow strange, on other devices I know they are either static or set by hardware. 
The point here is egg & chicken problem : you first need to use default adress ```0x63``` to be able to communicate with the device, execute the command and then modify esphome config to use the new address.
Can't see how to set a new adress without updating i2c component adress and make it working. It would require some tests to confirm what is the exact 'behavior'.

I would suggest you, either to remove unused functions or  provide  them as action  (ex [here](https://esphome.io/components/sensor/mhz19.html?highlight=mhz19#mhz19-calibrate-zero-action)), and detail procedure about how to use `change_i2c_address`.


My 2cent's

**Review by sebcaps — APPROVED** [review]

File: esphome/components/pmwcs3/sensor.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    var = cg.new_Pvariable(action_id, template_arg, parent)
+    return var
+
+
+PMWCS3NEWI2CADDRESS_SCHEMA = cv.Schema(
+    {
+        cv.GenerateID(): cv.use_id(PMWCS3Component),
+        cv.Required(CONF_NEWADDRESS): cv.templatable(int),
+    }
+)
```
Comment (jesserockz) [inline esphome/components/pmwcs3/sensor.py]:
```suggestion
PMWCS3_NEW_I2C_ADDRESS_SCHEMA = cv.maybe_simple_value(
    {
        cv.GenerateID(): cv.use_id(PMWCS3Component),
        cv.Required(CONF_ADDRESS): cv.templatable(cv.i2c_address),
    },
    key=CONF_ADDRESS,
)
```

File: esphome/components/pmwcs3/sensor.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+
+@automation.register_action(
+    "pmwcs3.newi2caddress",
```
Comment (jesserockz) [inline esphome/components/pmwcs3/sensor.py]: Dont try make things shorter, its just harder to read
```suggestion
    "pmwcs3.new_i2c_address",
```

File: esphome/components/pmwcs3/sensor.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    PMWCS3NewI2cAddressAction,
+    PMWCS3NEWI2CADDRESS_SCHEMA,
```
Comment (jesserockz) [inline esphome/components/pmwcs3/sensor.py]:
```suggestion
    PMWCS3_NEW_I2C_ADDRESS_SCHEMA,
```

File: esphome/components/pmwcs3/sensor.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+)
+async def pmwcs3newi2caddress_to_code(config, action_id, template_arg, args):
+    parent = await cg.get_variable(config[CONF_ID])
+    var = cg.new_Pvariable(action_id, template_arg, parent)
+    newaddress_ = await cg.templatable(config[CONF_NEWADDRESS], args, int)
+    cg.add(var.set_newaddress(newaddress_))
```
Comment (jesserockz) [inline esphome/components/pmwcs3/sensor.py]:
```suggestion
    address = await cg.templatable(config[CONF_ADDRESS], args, int)
    cg.add(var.set_new_address(address))
```

File: esphome/components/pmwcs3/sensor.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+)
+PMWCS3NewI2cAddressAction = pmwcs3_ns.class_(
+    "PMWCS3NewI2cAddressAction", automation.Action
+)
+
+CONFIG_SCHEMA = (
+    cv.Schema(
+        {
+            cv.GenerateID(): cv.declare_id(PMWCS3Component),
+            cv.Optional(CONF_ADDRESS): cv.i2c_address,
```
Comment (jesserockz) [inline esphome/components/pmwcs3/sensor.py]: Not required as its included with `i2c.i2c_device_schema()`
```suggestion
```

File: esphome/components/pmwcs3/pmwcs3.h [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+ protected:
+  PMWCS3Component *parent_;
+};
+
+template<typename... Ts> class PMWCS3NewI2cAddressAction : public Action<Ts...> {
+ public:
+  PMWCS3NewI2cAddressAction(PMWCS3Component *parent) : parent_(parent) {}
+  TEMPLATABLE_VALUE(int, newaddress)
+
+  void play(Ts... x) override { this->parent_->new_i2c_address(this->newaddress_.value(x...)); }
```
Comment (jesserockz) [inline esphome/components/pmwcs3/pmwcs3.h]:
```suggestion
  TEMPLATABLE_VALUE(int, new_address)

  void play(Ts... x) override { this->parent_->new_i2c_address(this->new_address_.value(x...)); }
```

File: esphome/components/pmwcs3/pmwcs3.h [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+  void set_vwc_sensor(sensor::Sensor *vwc_sensor) { vwc_sensor_ = vwc_sensor; }
+
+  void new_i2c_address(uint8_t newaddress);
+  void air_calibration();
+  void water_calibration();
+
+ protected:
+  void read_data_();
+  uint8_t newaddress_ = 0x63;
+  void set_newaddress_(uint8_t newaddress);
```
Comment (jesserockz) [inline esphome/components/pmwcs3/pmwcs3.h]: Not used
```suggestion
```

File: esphome/components/pmwcs3/pmwcs3.cpp [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+static const char *const TAG = "pmwcs3";
+
+void PMWCS3Component::new_i2c_address(uint8_t newaddress) {
+  if (!this->write_byte(PMWCS3_SET_I2C_ADDRESS, newaddress)) {
+    this->status_set_warning();
+    ESP_LOGW(TAG, "couldn't write the new I2C address %d", newaddress);
+    return;
+  }
+  ESP_LOGVV(TAG, "changed I2C address to %d", newaddress);
+}
```
Comment (jesserockz) [inline esphome/components/pmwcs3/pmwcs3.cpp]:
```suggestion
void PMWCS3Component::new_i2c_address(uint8_t address) {
  if (!this->write_byte(PMWCS3_SET_I2C_ADDRESS, address)) {
    this->status_set_warning();
    ESP_LOGW(TAG, "couldn't write the new I2C address %d", address);
    return;
  }
  this->set_i2c_address(address);  // Allows device to continue working until new firmware is written with new address.
  ESP_LOGVV(TAG, "changed I2C address to %d", address);
  this->status_clear_warning();
}
```

File: esphome/components/pmwcs3/sensor.py [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+    STATE_CLASS_MEASUREMENT,
+    ICON_THERMOMETER,
+)
+
+CODEOWNERS = ["@SeByDocKy"]
+DEPENDENCIES = ["i2c"]
+
+CONF_E25 = "e25"
+CONF_VWC = "vwc"
+CONF_NEWADDRESS = "newaddress"
```
Comment (jesserockz) [inline esphome/components/pmwcs3/sensor.py]:
```suggestion
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Review by jesserockz — DISMISSED** [review]

**Review by jesserockz — APPROVED** [review]

**Pull request merged**